File: src/main/scala/firefly/Unification.scala
package firefly
import firefly.Firefly_Core._

import firefly.Syntax_._
object Unification_ {

case class Unification(var substitution : Firefly_Core.Map[Firefly_Core.Int, Syntax_.Type], var constraints : Firefly_Core.Map[Firefly_Core.Int, Firefly_Core.Map[Firefly_Core.String, Unification_.ConstraintGenerics]], var nextTypeVariableIndex : Firefly_Core.Int, instances : Firefly_Core.Map[Unification_.InstanceKey, Unification_.InstanceValue])

case class ConstraintGenerics(at : Syntax_.Location, generics : Firefly_Core.List[Syntax_.Type])

case class InstanceKey(traitName : Firefly_Core.String, typeName : Firefly_Core.String)

case class InstanceValue(generics : Firefly_Core.List[Firefly_Core.String], constraints : Firefly_Core.List[Syntax_.Constraint], traitType : Syntax_.Type)

def make(instances : Firefly_Core.List[Syntax_.DInstance]) : Unification_.Unification = {
def fail[T](at : Syntax_.Location, message : Firefly_Core.String) : T = {
Firefly_Core.panic(((message + " ") + at.show()))
}
Unification_.Unification(Firefly_Core.Map(), Firefly_Core.Map(), 2, instances.map({(definition) =>
pipe_dot(definition.traitType)({
case (Syntax_.TConstructor(at, name, List(Syntax_.TConstructor(_, typeName, _), _ @ _*))) =>
Firefly_Core.Pair(Unification_.InstanceKey(name, typeName), Unification_.InstanceValue(generics = definition.generics, constraints = definition.constraints, traitType = definition.traitType))
case (Syntax_.TConstructor(at, name, _)) =>
fail(at, (("Instance requires type arguments: " + name) + "[]"))
case (Syntax_.TVariable(at, i)) =>
fail(at, ("Unexpected type variable: $" + i))
})
}).toMap)
}
implicit class Unification_extend0(self : Unification_.Unification) {

def fail[T](at : Syntax_.Location, message : Firefly_Core.String) : T = {
Firefly_Core.panic(((message + " ") + at.show()))
}

def freshTypeVariable(at : Syntax_.Location) : Syntax_.Type = {
val result = Syntax_.TVariable(at, self.nextTypeVariableIndex);
self.nextTypeVariableIndex += 2;
result
}

def instantiate(instantiation : Firefly_Core.Map[Firefly_Core.String, Syntax_.Type], type_ : Syntax_.Type) : Syntax_.Type = (instantiation, type_) match {
case (_, Syntax_.TConstructor(at, name, List())) =>
pipe_dot(instantiation.get(name))({
case (Firefly_Core.Some(t)) =>
t
case (Firefly_Core.None()) =>
type_
})
case (_, Syntax_.TConstructor(at, name, generics)) =>
Syntax_.TConstructor(at, name, generics.map({(_w1) =>
self.instantiate(instantiation, _w1)
}))
case (_, Syntax_.TVariable(_, i)) if self.has(i) =>
self.instantiate(instantiation, self.get(i))
case (_, Syntax_.TVariable(_, i)) =>
type_
}

def constrain(at : Syntax_.Location, type_ : Syntax_.Type, constraintName : Firefly_Core.String, generics : Firefly_Core.List[Syntax_.Type]) : Firefly_Core.Unit = {
pipe_dot(type_)({
case (Syntax_.TVariable(_, i)) if self.has(i) =>
self.constrain(at, self.get(i), constraintName, generics)
case (Syntax_.TVariable(_, i)) =>
pipe_dot(self.constraints.get(i))({
case (Firefly_Core.None()) =>
self.constraints += Firefly_Core.Pair(i, Firefly_Core.Map(Firefly_Core.Pair(constraintName, Unification_.ConstraintGenerics(at, generics))))
case (Firefly_Core.Some(map)) =>
pipe_dot(map.get(constraintName))({
case (Firefly_Core.None()) =>
val newMap = map.updated(constraintName, Unification_.ConstraintGenerics(at, generics));
self.constraints = self.constraints.updated(i, newMap)
case (Firefly_Core.Some(Unification_.ConstraintGenerics(_, generics2))) =>
generics.zip(generics2).each({
case (Firefly_Core.Pair(t1, t2)) =>
self.unify(at, t1, t2)
})
})
})
case (Syntax_.TConstructor(_, name, generics2)) =>
pipe_dot(self.instances.get(Unification_.InstanceKey(constraintName, name)))({
case (Firefly_Core.None()) =>
val g1 = Firefly_Core.if_(generics.getEmpty(), {() =>
""
}).else_({() =>
"[...]"
});
val g2 = Firefly_Core.if_(generics2.getEmpty(), {() =>
""
}).else_({() =>
"[...]"
});
self.fail(at, ((((("No such instance: " + name) + g2) + ": ") + constraintName) + g1))
case (Firefly_Core.Some(definition)) =>
val typeVariables = definition.generics.map({(_) =>
self.freshTypeVariable(at)
});
val instantiation = definition.generics.zip(typeVariables).toMap;
val traitType1 = self.instantiate(instantiation, definition.traitType);
val traitType2 = Syntax_.TConstructor(at, constraintName, (List(List(type_), generics).flatten));
self.unify(at, traitType1, traitType2);
definition.constraints.each({(constraint) =>
pipe_dot(self.instantiate(instantiation, constraint.representation))({
case (Syntax_.TConstructor(_, newConstraintName, newGenerics)) =>
self.constrain(at, type_, newConstraintName, newGenerics)
case (Syntax_.TVariable(_, i)) =>
self.fail(at, ("Constraint can't be a type variable: $" + i))
})
})
})
})
}

def get(index : Firefly_Core.Int) : Syntax_.Type = {
pipe_dot(self.substitution(index))({
case (Syntax_.TVariable(_, i)) if self.has(i) =>
val t = self.get(i);
self.substitution += Firefly_Core.Pair(index, t);
t
case (t) =>
t
})
}

def has(index : Firefly_Core.Int) : Firefly_Core.Bool = {
self.substitution.contains(index)
}

def substitute(type_ : Syntax_.Type) : Syntax_.Type = (type_) match {
case (Syntax_.TVariable(_, i)) =>
Firefly_Core.if_(self.has(i), {() =>
self.substitute(self.get(i))
}).else_({() =>
type_
})
case (t : Syntax_.TConstructor) =>
t.copy(generics = t.generics.map({(t) =>
self.substitute(t)
}))
}

def unify(at : Syntax_.Location, t1 : Syntax_.Type, t2 : Syntax_.Type) : Firefly_Core.Unit = (at, t1, t2) match {
case (_, Syntax_.TVariable(_, i1), Syntax_.TVariable(_, i2)) if (i1 == i2) =>
Firefly_Core.Unit()
case (_, Syntax_.TVariable(_, i), _) if self.has(i) =>
self.unify(at, self.get(i), t2)
case (_, _, Syntax_.TVariable(_, i)) if self.has(i) =>
self.unify(at, t1, self.get(i))
case (_, Syntax_.TVariable(_, i), _) =>
self.bind(at, i, t2)
case (_, _, Syntax_.TVariable(_, i)) =>
self.bind(at, i, t1)
case (_, Syntax_.TConstructor(_, name1, generics1), Syntax_.TConstructor(_, name2, generics2)) =>
Firefly_Core.if_(((name1 != name2) || (generics1.getSize() != generics2.getSize())), {() =>
self.fail(at, ((("Type mismatch: " + self.substitute(t1).show()) + " vs. ") + self.substitute(t2).show()))
});
generics1.zip(generics2).each({
case (Firefly_Core.Pair(t1, t2)) =>
self.unify(at, t1, t2)
})
}

def bind(at : Syntax_.Location, index : Firefly_Core.Int, type_ : Syntax_.Type) : Firefly_Core.Unit = {
Firefly_Core.if_(self.occursIn(index, type_), {() =>
self.fail(at, ((("Infinite type: $" + index) + " = ") + self.substitute(type_).show()))
});
self.substitution += Firefly_Core.Pair(index, type_);
self.constraints.get(index).each({(map) =>
self.constraints -= index;
map.pairs().each({
case (Firefly_Core.Pair(name, Unification_.ConstraintGenerics(at2, generics))) =>
self.constrain(at2, type_, name, generics)
})
})
}

def occursIn(index : Firefly_Core.Int, t : Syntax_.Type) : Firefly_Core.Bool = (index, t) match {
case (_, Syntax_.TVariable(_, i)) if self.has(i) =>
self.occursIn(index, self.get(i))
case (_, Syntax_.TVariable(_, i)) =>
(i == index)
case (_, Syntax_.TConstructor(_, _, generics)) =>
generics.any({(t) =>
self.occursIn(index, t)
})
}

}


}

File: src/main/js/ff/compiler/Unification.mjs
import * as ff_compiler_Unification from "../../ff/compiler/Unification.mjs"

import * as ff_compiler_Syntax from "../../ff/compiler/Syntax.mjs"

import * as ff_core_Array from "../../ff/core/Array.mjs"

import * as ff_core_ArrayBuilder from "../../ff/core/ArrayBuilder.mjs"

import * as ff_core_Bool from "../../ff/core/Bool.mjs"

import * as ff_core_Char from "../../ff/core/Char.mjs"

import * as ff_core_Core from "../../ff/core/Core.mjs"

import * as ff_core_FileSystem from "../../ff/core/FileSystem.mjs"

import * as ff_core_Int from "../../ff/core/Int.mjs"

import * as ff_core_List from "../../ff/core/List.mjs"

import * as ff_core_Log from "../../ff/core/Log.mjs"

import * as ff_core_Map from "../../ff/core/Map.mjs"

import * as ff_core_Nothing from "../../ff/core/Nothing.mjs"

import * as ff_core_Option from "../../ff/core/Option.mjs"

import * as ff_core_Pair from "../../ff/core/Pair.mjs"

import * as ff_core_Set from "../../ff/core/Set.mjs"

import * as ff_core_String from "../../ff/core/String.mjs"

import * as ff_core_System from "../../ff/core/System.mjs"

import * as ff_core_Try from "../../ff/core/Try.mjs"

import * as ff_core_Unit from "../../ff/core/Unit.mjs"

// type Unification
export function Unification(substitution_, constraints_, nextTypeVariableIndex_, instances_) {
return {_: 'Unification', substitution_, constraints_, nextTypeVariableIndex_, instances_};
}

// type ConstraintGenerics
export function ConstraintGenerics(at_, generics_) {
return {_: 'ConstraintGenerics', at_, generics_};
}

// type InstanceKey
export function InstanceKey(traitName_, typeName_) {
return {_: 'InstanceKey', traitName_, typeName_};
}

// type InstanceValue
export function InstanceValue(generics_, constraints_, traitType_) {
return {_: 'InstanceValue', generics_, constraints_, traitType_};
}



export function make_(instances_) {
function fail_(at_, message_) {
return ff_core_Core.panic_(((message_ + " ") + ff_compiler_Syntax.Location_show(at_)))
}
return ff_compiler_Unification.Unification(ff_core_Map.empty_(), ff_core_Map.empty_(), 2, ff_core_List.List_toMap(ff_core_List.List_map(instances_, ((definition_) => {
{
const _1 = definition_.traitType_
{
if(_1._ === 'TConstructor') {
const at_ = _1.at_
const name_ = _1.name_
if(_1.generics_._ === 'Link') {
if(_1.generics_.head_._ === 'TConstructor') {
const typeName_ = _1.generics_.head_.name_
return ff_core_Pair.Pair(ff_compiler_Unification.InstanceKey(name_, typeName_), ff_compiler_Unification.InstanceValue(definition_.generics_, definition_.constraints_, definition_.traitType_))
return
}}}
}
{
if(_1._ === 'TConstructor') {
const at_ = _1.at_
const name_ = _1.name_
return fail_(at_, (("Instance requires type arguments: " + name_) + "[]"))
return
}
}
{
if(_1._ === 'TVariable') {
const at_ = _1.at_
const i_ = _1.index_
return fail_(at_, ("Unexpected type variable: $" + i_))
return
}
}
throw new Error('Unexhaustive pattern match')
}
}))))
}

export function Unification_fail(self_, at_, message_) {
return ff_core_Core.panic_(((message_ + " ") + ff_compiler_Syntax.Location_show(at_)))
}

export function Unification_freshTypeVariable(self_, at_) {
const result_ = ff_compiler_Syntax.TVariable(at_, self_.nextTypeVariableIndex_)
self_.nextTypeVariableIndex_ += 2
return result_
}

export function Unification_instantiate(self_, instantiation_, type_){
const self_a = self_
const instantiation_a = instantiation_
const type_a = type_
{
const self_ = self_a
if(type_a._ === 'TConstructor') {
const at_ = type_a.at_
const name_ = type_a.name_
if(type_a.generics_._ === 'Empty') {
{
const _1 = ff_core_Map.Map_get(instantiation_, name_)
{
if(_1._ === 'Some') {
const t_ = _1.value_
return t_
return
}
}
{
if(_1._ === 'None') {
return type_
return
}
}
throw new Error('Unexhaustive pattern match')
}
return
}}
}
{
const self_ = self_a
if(type_a._ === 'TConstructor') {
const at_ = type_a.at_
const name_ = type_a.name_
const generics_ = type_a.generics_
return ff_compiler_Syntax.TConstructor(at_, name_, ff_core_List.List_map(generics_, ((_w1) => {
return ff_compiler_Unification.Unification_instantiate(self_, instantiation_, _w1)
})))
return
}
}
{
const self_ = self_a
if(type_a._ === 'TVariable') {
const i_ = type_a.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_instantiate(self_, instantiation_, ff_compiler_Unification.Unification_get(self_, i_))
return
}}
}
{
const self_ = self_a
if(type_a._ === 'TVariable') {
const i_ = type_a.index_
return type_
return
}
}
throw new Error('Unexhaustive pattern match')
}

export function Unification_constrain(self_, at_, type_, constraintName_, generics_) {
{
const _1 = type_
{
if(_1._ === 'TVariable') {
const i_ = _1.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_constrain(self_, at_, ff_compiler_Unification.Unification_get(self_, i_), constraintName_, generics_)
return
}}
}
{
if(_1._ === 'TVariable') {
const i_ = _1.index_
{
const _1 = ff_core_Map.Map_get(self_.constraints_, i_)
{
if(_1._ === 'None') {
self_.constraints_ = ff_core_Map.Map_add(self_.constraints_, i_, ff_core_List.List_toMap(ff_core_Array.Array_toList([ff_core_Pair.Pair(constraintName_, ff_compiler_Unification.ConstraintGenerics(at_, generics_))])))
return
}
}
{
if(_1._ === 'Some') {
const map_ = _1.value_
{
const _1 = ff_core_Map.Map_get(map_, constraintName_)
{
if(_1._ === 'None') {
const newMap_ = ff_core_Map.Map_add(map_, constraintName_, ff_compiler_Unification.ConstraintGenerics(at_, generics_))
self_.constraints_ = ff_core_Map.Map_add(self_.constraints_, i_, newMap_)
return
}
}
{
if(_1._ === 'Some') {
const generics2_ = _1.value_.generics_
return ff_core_List.List_each(ff_core_List.List_zip(generics_, generics2_), ((_1) => {
{
const t1_ = _1.first_
const t2_ = _1.second_
return ff_compiler_Unification.Unification_unify(self_, at_, t1_, t2_)
return

}
throw new Error('Unexhaustive pattern match')
}))
return
}
}
throw new Error('Unexhaustive pattern match')
}
return
}
}
throw new Error('Unexhaustive pattern match')
}
return
}
}
{
if(_1._ === 'TConstructor') {
const name_ = _1.name_
const generics2_ = _1.generics_
{
const _1 = ff_core_Map.Map_get(self_.instances_, ff_compiler_Unification.InstanceKey(constraintName_, name_))
{
if(_1._ === 'None') {
const g1_ = (ff_core_List.List_isEmpty(generics_)
? ""
: "[...]")
const g2_ = (ff_core_List.List_isEmpty(generics2_)
? ""
: "[...]")
return ff_compiler_Unification.Unification_fail(self_, at_, ((((("No such instance: " + name_) + g2_) + ": ") + constraintName_) + g1_))
return
}
}
{
if(_1._ === 'Some') {
const definition_ = _1.value_
const typeVariables_ = ff_core_List.List_map(definition_.generics_, ((_) => {
return ff_compiler_Unification.Unification_freshTypeVariable(self_, at_)
}))
const instantiation_ = ff_core_List.List_toMap(ff_core_List.List_zip(definition_.generics_, typeVariables_))
const traitType1_ = ff_compiler_Unification.Unification_instantiate(self_, instantiation_, definition_.traitType_)
const traitType2_ = ff_compiler_Syntax.TConstructor(at_, constraintName_, ff_core_Array.Array_toList([type_, ...ff_core_List.List_toArray(generics_)]))
ff_compiler_Unification.Unification_unify(self_, at_, traitType1_, traitType2_)
return ff_core_List.List_each(definition_.constraints_, ((constraint_) => {
{
const _1 = ff_compiler_Unification.Unification_instantiate(self_, instantiation_, constraint_.representation_)
{
if(_1._ === 'TConstructor') {
const newConstraintName_ = _1.name_
const newGenerics_ = _1.generics_
return ff_compiler_Unification.Unification_constrain(self_, at_, type_, newConstraintName_, newGenerics_)
return
}
}
{
if(_1._ === 'TVariable') {
const i_ = _1.index_
return ff_compiler_Unification.Unification_fail(self_, at_, ("Constraint can't be a type variable: $" + i_))
return
}
}
throw new Error('Unexhaustive pattern match')
}
}))
return
}
}
throw new Error('Unexhaustive pattern match')
}
return
}
}
throw new Error('Unexhaustive pattern match')
}
}

export function Unification_get(self_, index_) {
{
const _1 = ff_core_Map.Map_expect(self_.substitution_, index_)
{
if(_1._ === 'TVariable') {
const i_ = _1.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
const t_ = ff_compiler_Unification.Unification_get(self_, i_)
self_.substitution_ = ff_core_Map.Map_add(self_.substitution_, index_, t_)
return t_
return
}}
}
{
const t_ = _1
return t_
return

}
throw new Error('Unexhaustive pattern match')
}
}

export function Unification_has(self_, index_) {
return ff_core_Map.Map_contains(self_.substitution_, index_)
}

export function Unification_substitute(self_, type_){
const self_a = self_
const type_a = type_
{
const self_ = self_a
if(type_a._ === 'TVariable') {
const i_ = type_a.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_substitute(self_, ff_compiler_Unification.Unification_get(self_, i_))
} else {
return type_
}
return
}
}
{
const self_ = self_a
if(type_a._ === 'TConstructor') {
const t_ = type_a
{
const _1 = t_
{
const _c = _1
return ff_compiler_Syntax.TConstructor(_c.at_, _c.name_, ff_core_List.List_map(t_.generics_, ((t_) => {
return ff_compiler_Unification.Unification_substitute(self_, t_)
})))
return

}
throw new Error('Unexhaustive pattern match')
}
return
}
}
throw new Error('Unexhaustive pattern match')
}

export function Unification_unify(self_, at_, t1_, t2_){
const self_a = self_
const at_a = at_
const t1_a = t1_
const t2_a = t2_
{
const self_ = self_a
if(t1_a._ === 'TVariable') {
const i1_ = t1_a.index_
if(t2_a._ === 'TVariable') {
const i2_ = t2_a.index_
if((i1_ == i2_)) {

return
}}}
}
{
const self_ = self_a
if(t1_a._ === 'TVariable') {
const i_ = t1_a.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_unify(self_, at_, ff_compiler_Unification.Unification_get(self_, i_), t2_)
return
}}
}
{
const self_ = self_a
if(t2_a._ === 'TVariable') {
const i_ = t2_a.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_unify(self_, at_, t1_, ff_compiler_Unification.Unification_get(self_, i_))
return
}}
}
{
const self_ = self_a
if(t1_a._ === 'TVariable') {
const i_ = t1_a.index_
return ff_compiler_Unification.Unification_bind(self_, at_, i_, t2_)
return
}
}
{
const self_ = self_a
if(t2_a._ === 'TVariable') {
const i_ = t2_a.index_
return ff_compiler_Unification.Unification_bind(self_, at_, i_, t1_)
return
}
}
{
const self_ = self_a
if(t1_a._ === 'TConstructor') {
const name1_ = t1_a.name_
const generics1_ = t1_a.generics_
if(t2_a._ === 'TConstructor') {
const name2_ = t2_a.name_
const generics2_ = t2_a.generics_
if(((name1_ != name2_) || (ff_core_List.List_size(generics1_) != ff_core_List.List_size(generics2_)))) {
ff_compiler_Unification.Unification_fail(self_, at_, ((("Type mismatch: " + ff_compiler_Syntax.Type_show(ff_compiler_Unification.Unification_substitute(self_, t1_))) + " vs. ") + ff_compiler_Syntax.Type_show(ff_compiler_Unification.Unification_substitute(self_, t2_))))
}
return ff_core_List.List_each(ff_core_List.List_zip(generics1_, generics2_), ((_1) => {
{
const t1_ = _1.first_
const t2_ = _1.second_
return ff_compiler_Unification.Unification_unify(self_, at_, t1_, t2_)
return

}
throw new Error('Unexhaustive pattern match')
}))
return
}}
}
throw new Error('Unexhaustive pattern match')
}

export function Unification_bind(self_, at_, index_, type_) {
if(ff_compiler_Unification.Unification_occursIn(self_, index_, type_)) {
ff_compiler_Unification.Unification_fail(self_, at_, ((("Infinite type: $" + index_) + " = ") + ff_compiler_Syntax.Type_show(ff_compiler_Unification.Unification_substitute(self_, type_))))
}
self_.substitution_ = ff_core_Map.Map_add(self_.substitution_, index_, type_)
return ff_core_Option.Option_each(ff_core_Map.Map_get(self_.constraints_, index_), ((map_) => {
self_.constraints_ = ff_core_Map.Map_remove(self_.constraints_, index_)
return ff_core_List.List_each(ff_core_Map.Map_pairs(map_), ((_1) => {
{
const name_ = _1.first_
const at2_ = _1.second_.at_
const generics_ = _1.second_.generics_
return ff_compiler_Unification.Unification_constrain(self_, at2_, type_, name_, generics_)
return

}
throw new Error('Unexhaustive pattern match')
}))
}))
}

export function Unification_occursIn(self_, index_, t_){
const self_a = self_
const index_a = index_
const t_a = t_
{
const self_ = self_a
if(t_a._ === 'TVariable') {
const i_ = t_a.index_
if(ff_compiler_Unification.Unification_has(self_, i_)) {
return ff_compiler_Unification.Unification_occursIn(self_, index_, ff_compiler_Unification.Unification_get(self_, i_))
return
}}
}
{
const self_ = self_a
if(t_a._ === 'TVariable') {
const i_ = t_a.index_
return (i_ == index_)
return
}
}
{
const self_ = self_a
if(t_a._ === 'TConstructor') {
const generics_ = t_a.generics_
return ff_core_List.List_any(generics_, ((t_) => {
return ff_compiler_Unification.Unification_occursIn(self_, index_, t_)
}))
return
}
}
throw new Error('Unexhaustive pattern match')
}


